src/substrait/function_registry.py

```diff
@@ -0,0 +1,316 @@
+from substrait.gen.proto.parameterized_types_pb2 import ParameterizedType
+from substrait.gen.proto.type_pb2 import Type
+from importlib.resources import files as importlib_files
+import itertools
+from collections import defaultdict
+from collections.abc import Mapping
+from pathlib import Path
+from typing import Any, Optional, Union
+from .derivation_expression import evaluate
+
+import yaml
+import re
+
+_normalized_key_names = {
+    "binary": "vbin",
+    "interval_compound": "icompound",
+    "interval_day": "iday",
+    "interval_year": "iyear",
+    "string": "str",
+    "timestamp": "ts",
+    "timestamp_tz": "tstz",
+}
+
+
+def normalize_substrait_type_names(typ: str) -> str:
+    # First strip off any punctuation
+    typ = typ.strip("?").lower()
+
+    # Common prefixes whose information does not matter to an extension function
+    # signature
+    for complex_type, abbr in [
+        ("fixedchar", "fchar"),
+        ("varchar", "vchar"),
+        ("fixedbinary", "fbin"),
+        ("decimal", "dec"),
+        ("precision_timestamp", "pts"),
+        ("precision_timestamp_tz", "ptstz"),
+        ("struct", "struct"),
+        ("list", "list"),
+        ("map", "map"),
+        ("any", "any"),
+        ("boolean", "bool"),
+    ]:
+        if typ.lower().startswith(complex_type):
+            typ = abbr
+
+    # Then pass through the dictionary of mappings, defaulting to just the
+    # existing string
+    typ = _normalized_key_names.get(typ.lower(), typ.lower())
+    return typ
+
+
+id_generator = itertools.count(1)
```

Comment:
For long running programs this id generation would get pretty big.  Would it be better to reset for each plan (perhaps requiring a function registry class)?

Reply:
oops, my bad. this should not be here. `id_generator` is also being created in `FunctionRegistry` constructor, that's the one that's supposed to be used instead. That way it will be up to the application itself whether it wants to reuse a single FunctionRegistry (shared id generation) or create a brand new one for each plan.